teacher_student_dynamics/networks/network_ensembles/base_ensemble.py

```diff
@@ -85,6 +85,20 @@ def _init_network(self, num_heads: int = 1):
             freeze=True,
         )
 
+    def project_networks(self, projections: List[List[torch.Tensor]]) -> None:
+        with torch.no_grad():
+
+            for i in range(self._ensemble_size):
+
+                # MxD
+                unprojected_weights = self._networks[i].layers[0].weight.data
+                # DxD
+                projection = projections[i]
+
+                self._networks[i].layers[0].weight.data = unprojected_weights.mm(
+                    projection.T
+                ).mm(projection)
+
```

Comment:
I might be imagining the projection on lines 98 to 100 wrong but:
1. Is unprojected_weights (call it W) an N x D matrix (so you multiply weights on the left of data when forward propping)?
2. If this is the case then each row of W is a feature for one hidden neuron. Projection then is the D x D matrix of eigenvectors in the columns (so F_tilde)? So W @ F_tilde would give the alignment of each feature (W row) with each eigenvec (F_tilde column). But lines 98 and 99 seem to be using W @ F_tilde.T
3. W @ F_tilde would then give coefficients along the row for the contribution we want from each F_tilde column (with 0 on dimensions of F_tilde we don't want). So (W @ F_tilde) @ F_tilde.T should give what we want no? New weights reconstructed as a linear combo of rows of F_tilde.T (columns of the eigenvecs)
4. So long way of saying, should the transpose of the two multiplications by the projection not be swopped?

Reply:
yes.